File: app/cfpb_complaint.py
"""
CFPB Complaint Generator
Generate properly formatted complaints for the Consumer Financial Protection Bureau.

The CFPB accepts complaints about credit reporting issues and requires specific information.
This module generates complaint text optimized for CFPB submission.
"""

import streamlit as st
from datetime import datetime
from typing import Dict, List, Any, Optional
from dataclasses import dataclass


@dataclass
class CFPBComplaint:
    """Structured CFPB complaint data."""
    product: str  # "Credit reporting, credit repair services, or other personal consumer reports"
    sub_product: str  # "Credit reporting"
    issue: str  # e.g., "Incorrect information on your report"
    sub_issue: str  # e.g., "Account status incorrect"
    company_name: str
    complaint_narrative: str
    desired_resolution: str
    consumer_consent: bool = True


# CFPB complaint categories for credit reporting
CFPB_ISSUES = {
    "Incorrect information on your report": [
        "Account status incorrect",
        "Account information incorrect",
        "Personal information incorrect",
        "Public record information inaccurate",
        "Information belongs to someone else",
        "Old information reappears or never goes away"
    ],
    "Problem with a credit reporting company's investigation": [
        "Their investigation did not fix an error on your report",
        "Investigation took more than 30 days",
        "Was not notified of investigation status or results",
        "Difficulty submitting a dispute or getting information about a dispute"
    ],
    "Improper use of your report": [
        "Reporting company used your report improperly",
        "Credit inquiries on your report that you don't recognize"
    ],
    "Problem with a company's investigation into an existing problem": [
        "Their investigation did not fix an error on your report",
        "Investigation took more than 30 days",
        "Was not notified of investigation status or results"
    ],
    "Unable to get your credit report or credit score": [
        "Problem getting your free annual credit report",
        "Other problem getting your credit report or credit score"
    ]
}


def generate_complaint_narrative(
    flags: List[Dict],
    fields: Dict[str, Any],
    consumer_info: Dict[str, str],
    timeline_data: List[Dict] = None
) -> str:
    """Generate a compelling complaint narrative from detected issues."""

    creditor = str(fields.get('creditor_name') or fields.get('furnisher_or_collector') or 'the furnisher')
    account_type = str(fields.get('account_type') or 'account')
    dofd = str(fields.get('dofd') or 'unknown')
    removal_date = str(fields.get('estimated_removal_date') or 'unknown')

    narrative = []

    # Opening
    narrative.append(f"I am filing this complaint regarding inaccurate information being reported on my credit report by {creditor}.")
    narrative.append("")

    # Account identification
    narrative.append("ACCOUNT INFORMATION:")
    narrative.append(f"- Creditor/Furnisher: {creditor}")
    if fields.get('account_number'):
        # Mask account number for privacy
        acct = fields.get('account_number', '')
        masked = 'XXXX' + acct[-4:] if len(acct) > 4 else acct
        narrative.append(f"- Account Number (last 4): {masked}")
    narrative.append(f"- Account Type: {account_type}")
    narrative.append("")

    # Specific violations found
    narrative.append("SPECIFIC ISSUES IDENTIFIED:")
    narrative.append("")

    high_flags = [f for f in flags if f.get('severity') == 'high']
    medium_flags = [f for f in flags if f.get('severity') == 'medium']

    for i, flag in enumerate(high_flags + medium_flags, 1):
        narrative.append(f"{i}. {flag.get('rule_name', 'Issue')}")
        narrative.append(f"   {flag.get('explanation', '')}")
        if flag.get('why_it_matters'):
            narrative.append(f"   Impact: {flag.get('why_it_matters', '')}")
        narrative.append("")

    # Timeline evidence if available
    if timeline_data and len(timeline_data) >= 2:
        narrative.append("TIMELINE EVIDENCE OF MANIPULATION:")
        narrative.append("")
        for entry in sorted(timeline_data, key=lambda x: x.get('report_date', '')):
            narrative.append(f"- Report dated {entry.get('report_date')}:")
            if entry.get('dofd'):
                narrative.append(f"  DOFD shown: {entry.get('dofd')}")
            if entry.get('date_opened'):
                narrative.append(f"  Date Opened: {entry.get('date_opened')}")
            if entry.get('removal_date'):
                narrative.append(f"  Removal Date: {entry.get('removal_date')}")
        narrative.append("")
        narrative.append("The dates changed between reports, which is evidence of illegal re-aging under the FCRA.")
        narrative.append("")

    # Legal basis
    narrative.append("LEGAL BASIS:")
    narrative.append("")
    narrative.append("Under the Fair Credit Reporting Act (FCRA), 15 U.S.C. § 1681:")
    narrative.append("- Section 1681e(b) requires credit reporting agencies to follow reasonable procedures to assure maximum possible accuracy")
    narrative.append("- Section 1681s-2(a) requires furnishers to report accurate information")
    narrative.append("- Section 1681c(a) limits reporting of most negative items to 7 years from the date of first delinquency")
    narrative.append("")

    if any('re-ag' in str(f.get('explanation', '')).lower() for f in flags):
        narrative.append("The practice of 're-aging' debt by manipulating dates is a violation of these provisions and constitutes willful non-compliance with the FCRA.")
        narrative.append("")

    # Previous dispute attempts
    narrative.append("PREVIOUS ATTEMPTS TO RESOLVE:")
    narrative.append("")
    narrative.append("I have attempted to resolve this matter directly with the credit bureaus and/or the furnisher through the standard dispute process. However, the inaccurate information continues to appear on my credit report.")
    narrative.append("")

    # Harm suffered
    narrative.append("HARM SUFFERED:")
    narrative.append("")
    narrative.append("This inaccurate information has negatively impacted my credit score and my ability to obtain credit, housing, and/or employment. The continued reporting of this inaccurate information causes ongoing harm to my financial well-being.")
    narrative.append("")

    return "\n".join(narrative)


def generate_desired_resolution(flags: List[Dict], fields: Dict) -> str:
    """Generate appropriate resolution request."""

    resolutions = []

    # Always request deletion or correction
    resolutions.append("1. IMMEDIATE DELETION of the inaccurate tradeline from all three credit bureaus (Experian, Equifax, TransUnion)")

    # If re-aging detected
    if any('re-ag' in str(f.get('explanation', '')).lower() for f in flags):
        resolutions.append("2. Investigation into the furnisher's reporting practices for evidence of systematic re-aging")

    # Request investigation
    resolutions.append("3. A thorough investigation into the accuracy of the reported information")

    # Request documentation
    resolutions.append("4. Copies of any documentation the furnisher claims supports the accuracy of this information")

    # Statutory damages mention
    if any(f.get('severity') == 'high' for f in flags):
        resolutions.append("5. If willful non-compliance is determined, I reserve my right to pursue statutory damages under 15 U.S.C. § 1681n")

    return "\n".join(resolutions)


def determine_cfpb_issue(flags: List[Dict]) -> tuple:
    """Determine the best CFPB issue category based on flags."""

    # Check for date manipulation (re-aging)
    date_issues = ['A1', 'A2', 'B1', 'B2', 'K6']
    has_date_issue = any(f.get('rule_id') in date_issues for f in flags)

    # Check for status issues
    status_issues = ['D1', 'D2', 'D3', 'E1', 'E2']
    has_status_issue = any(f.get('rule_id') in status_issues for f in flags)

    # Check for balance issues
    balance_issues = ['F1', 'F2', 'F3']
    has_balance_issue = any(f.get('rule_id') in balance_issues for f in flags)

    if has_date_issue:
        return ("Incorrect information on your report", "Old information reappears or never goes away")
    elif has_status_issue:
        return ("Incorrect information on your report", "Account status incorrect")
    elif has_balance_issue:
        return ("Incorrect information on your report", "Account information incorrect")
    else:
        return ("Incorrect information on your report", "Account information incorrect")


def render_cfpb_generator(st):
    """Render the CFPB Complaint Generator UI."""

    st.markdown("""
    <div style="margin-bottom: 20px;">
        <h2 style="color: #1e40af; margin-bottom: 8px;">CFPB Complaint Generator</h2>
        <p style="color: #64748b; font-size: 0.95rem;">
            Generate a properly formatted complaint for the Consumer Financial Protection Bureau.
        </p>
    </div>
    """, unsafe_allow_html=True)

    st.info("""
    **About CFPB Complaints:** The Consumer Financial Protection Bureau (CFPB) accepts complaints
    about credit reporting issues. Companies must respond within 15 days, and the CFPB tracks
    complaint patterns to identify bad actors. Filing a CFPB complaint creates an official record
    and often gets faster results than disputing directly with bureaus.
    """)

    # Check if we have case data
    flags = st.session_state.get('rule_flags', [])
    fields = st.session_state.get('editable_fields', {})

    if not flags:
        st.warning("""
        **No issues detected yet.**

        To generate a CFPB complaint, first analyze a credit report using "Check One Report" mode.
        Once issues are detected, return here to generate your complaint.
        """)

        # Manual entry option
        st.markdown("---")
        st.markdown("### Or Enter Information Manually")

        manual_creditor = st.text_input("Creditor/Collector Name", key="cfpb_creditor")
        manual_issue = st.text_area("Describe the issue", key="cfpb_issue", height=150,
                                    placeholder="Describe what's wrong with your credit report...")

        if manual_creditor and manual_issue:
            if st.button("Generate Basic Complaint", type="primary"):
                basic_complaint = f"""I am filing this complaint regarding inaccurate information being reported by {manual_creditor}.

ISSUE:
{manual_issue}

LEGAL BASIS:
Under the Fair Credit Reporting Act (FCRA), credit reporting agencies must follow reasonable procedures to assure maximum possible accuracy (15 U.S.C. § 1681e(b)), and furnishers must report accurate information (15 U.S.C. § 1681s-2(a)).

DESIRED RESOLUTION:
1. Immediate deletion or correction of the inaccurate information
2. Investigation into the accuracy of the reported information
3. Written confirmation of any changes made

I have attempted to resolve this matter through the standard dispute process without success."""

                st.markdown("### Your Complaint")
                st.text_area("Copy this text:", basic_complaint, height=400)

                st.markdown("""
                **Next Steps:**
                1. Go to [consumerfinance.gov/complaint](https://www.consumerfinance.gov/complaint/)
                2. Select "Credit reporting or other personal consumer reports"
                3. Follow the prompts and paste your complaint narrative
                """)
        return

    # We have flags - generate comprehensive complaint
    st.success(f"Found {len(flags)} issue(s) to include in your complaint.")

    # Get verified fields
    verified_fields = {k: v.get('value') if isinstance(v, dict) else v
                       for k, v in fields.items()}

    # Consumer info
    st.markdown("### Your Information")
    col1, col2 = st.columns(2)
    with col1:
        consumer_name = st.text_input("Your Name",
                                       value=st.session_state.get('consumer_info', {}).get('name', ''))
    with col2:
        consumer_state = st.text_input("Your State",
                                        value=st.session_state.get('consumer_info', {}).get('state', ''))

    # Timeline data if available
    timeline_data = st.session_state.get('timeline_entries', [])

    # Generate complaint
    issue, sub_issue = determine_cfpb_issue(flags)
    narrative = generate_complaint_narrative(flags, verified_fields,
                                              {'name': consumer_name, 'state': consumer_state},
                                              timeline_data if len(timeline_data) >= 2 else None)
    resolution = generate_desired_resolution(flags, verified_fields)

    st.markdown("---")
    st.markdown("### Generated Complaint")

    # Issue category
    st.markdown(f"**CFPB Category:** {issue}")
    st.markdown(f"**Sub-category:** {sub_issue}")

    # Narrative
    st.markdown("**Complaint Narrative:**")
    complaint_text = st.text_area("Your complaint (edit as needed):", narrative, height=500, key="cfpb_narrative")

    # Resolution
    st.markdown("**Desired Resolution:**")
    resolution_text = st.text_area("Desired outcome:", resolution, height=200, key="cfpb_resolution")

    # Download options
    st.markdown("---")
    st.markdown("### Save Your Complaint")

    col1, col2 = st.columns(2)
    with col1:
        full_complaint = f"""CFPB COMPLAINT
Generated: {datetime.now().strftime('%Y-%m-%d')}

CATEGORY: {issue}
SUB-CATEGORY: {sub_issue}

{'='*50}
COMPLAINT NARRATIVE:
{'='*50}

{complaint_text}

{'='*50}
DESIRED RESOLUTION:
{'='*50}

{resolution_text}
"""
        st.download_button(
            "Download as Text File",
            full_complaint,
            file_name=f"cfpb_complaint_{datetime.now().strftime('%Y%m%d')}.txt",
            mime="text/plain",
            use_container_width=True
        )

    with col2:
        if st.button("Copy to Clipboard", use_container_width=True):
            st.info("Use Ctrl+A then Ctrl+C in the text boxes above to copy")

    # Instructions
    st.markdown("---")
    st.markdown("### How to File Your CFPB Complaint")

    st.markdown("""
    <div style="background: #f0fdf4; border: 1px solid #bbf7d0; border-radius: 8px; padding: 20px; margin: 10px 0;">
        <h4 style="color: #166534; margin-top: 0;">Step-by-Step Instructions</h4>
        <ol style="color: #334155; line-height: 1.8;">
            <li>Go to <a href="https://www.consumerfinance.gov/complaint/" target="_blank"><strong>consumerfinance.gov/complaint</strong></a></li>
            <li>Select <strong>"Credit reporting or other personal consumer reports"</strong></li>
            <li>Select <strong>"Credit reporting"</strong></li>
            <li>Choose the issue category shown above</li>
            <li>Enter the company name (credit bureau or furnisher)</li>
            <li>Paste your complaint narrative in the description field</li>
            <li>Attach any supporting documents (credit report screenshots, dispute letters)</li>
            <li>Review and submit</li>
        </ol>
        <p style="color: #166534; font-weight: 600; margin-bottom: 0;">
            The company must respond within 15 days. You'll receive updates by email.
        </p>
    </div>
    """, unsafe_allow_html=True)

    # Tips
    with st.expander("Tips for Effective CFPB Complaints"):
        st.markdown("""
        **Do:**
        - Be specific about dates and account numbers
        - Include all rule violations detected by this tool
        - Attach copies of your credit report showing the errors
        - Mention any previous dispute attempts
        - State the specific harm you've suffered

        **Don't:**
        - Use emotional language or make threats
        - Include your full Social Security Number
        - Submit duplicate complaints about the same issue
        - Exaggerate or include unverified claims

        **Pro Tips:**
        - File separate complaints for each credit bureau if the same error appears on multiple reports
        - Keep records of all complaint reference numbers
        - Follow up if you don't hear back within 15 days
        - If the response is unsatisfactory, you can dispute it through the CFPB portal
        """)

File: pwa-app/src/lib/cfpb-complaint.ts
/**
 * CFPB Complaint Builder
 * Complete complaint generation with all required fields
 */

import { CreditFields, RuleFlag } from './rules';

export interface CFPBComplaint {
  // Product Information
  productType: 'credit_reporting' | 'debt_collection' | 'credit_card' | 'mortgage';
  subProduct?: string;

  // Issue Details
  issueType: string;
  subIssue?: string;

  // Company Information
  companyName: string;
  companyState?: string;

  // Consumer Information
  consumerName: string;
  consumerAddress: string;
  consumerCity: string;
  consumerState: string;
  consumerZip: string;
  consumerEmail?: string;
  consumerPhone?: string;

  // Complaint Details
  narrative: string;
  desiredResolution: string;

  // Additional Info
  submittedToCompany: boolean;
  companyResponse?: string;
  serviceMemberStatus: boolean;
  consentToPublish: 'yes' | 'no' | 'with_consent';

  // Attachments
  hasAttachments: boolean;
  attachmentDescriptions: string[];
}

export const CFPB_ISSUE_TYPES = {
  credit_reporting: [
    { value: 'incorrect_info', label: 'Incorrect information on your report', subIssues: [
      'Information belongs to someone else',
      'Account status incorrect',
      'Account information incorrect',
      'Personal information incorrect',
      'Public record information inaccurate'
    ]},
    { value: 'improper_use', label: 'Improper use of your report', subIssues: [
      'Reporting company used your report improperly',
      'Credit inquiries on your report that you don\'t recognize'
    ]},
    { value: 'unable_to_get_report', label: 'Unable to get your credit report or credit score', subIssues: [
      'Problem getting your free annual credit report',
      'Problem getting a credit report from a non-traditional source'
    ]},
    { value: 'problem_with_investigation', label: 'Problem with a credit reporting company\'s investigation', subIssues: [
      'Their investigation did not fix an error on your report',
      'Investigation took more than 30 days',
      'Was not notified of investigation status or results',
      'Difficulty submitting a dispute or getting information about a dispute over the phone'
    ]},
    { value: 'problem_with_fraud_alert', label: 'Problem with a company\'s investigation into an existing problem', subIssues: [
      'Difficulty getting a response from the company',
      'Difficulty getting a response from a credit bureau'
    ]}
  ],
  debt_collection: [
    { value: 'attempts_collect_not_owed', label: 'Attempts to collect debt not owed', subIssues: [
      'Debt was paid',
      'Debt was discharged in bankruptcy',
      'Debt was result of identity theft',
      'Debt is not yours'
    ]},
    { value: 'written_notification', label: 'Written notification about debt', subIssues: [
      'Didn\'t receive enough information to verify debt',
      'Didn\'t receive notice of rights when contacted'
    ]},
    { value: 'false_statements', label: 'False statements or representation', subIssues: [
      'Attempted to collect wrong amount',
      'Impersonated attorney, law enforcement, or government official',
      'Indicated you committed crime by not paying debt'
    ]},
    { value: 'communication_tactics', label: 'Communication tactics', subIssues: [
      'Frequent or repeated calls',
      'Called before 8am or after 9pm',
      'Used obscene, profane, or abusive language',
      'Contacted you after you asked them to stop'
    ]},
    { value: 'threatened_actions', label: 'Threatened to take negative or legal action', subIssues: [
      'Threatened to sue you for very old debt',
      'Threatened arrest or jail if you don\'t pay',
      'Threatened to take legal action'
    ]}
  ]
};

/**
 * Build a complete CFPB complaint from analysis data
 */
export function buildCFPBComplaint(
  fields: CreditFields,
  flags: RuleFlag[],
  consumerInfo: {
    name: string;
    address: string;
    city: string;
    state: string;
    zip: string;
    email?: string;
    phone?: string;
  }
): CFPBComplaint {
  // Determine product type
  const isCollection = (fields.accountType || '').toLowerCase().includes('collection') ||
                      (fields.furnisherOrCollector || '').toLowerCase().includes('collection');

  const productType = isCollection ? 'debt_collection' : 'credit_reporting';

  // Determine issue type based on flags
  const hasReaging = flags.some(f => ['B1', 'B2', 'B3', 'K6'].includes(f.ruleId));
  const hasIncorrectInfo = flags.some(f => ['D1', 'M1', 'M2'].includes(f.ruleId));
  const hasInvestigationIssue = flags.some(f => f.severity === 'high');

  let issueType = 'incorrect_info';
  let subIssue = 'Account information incorrect';

  if (hasReaging) {
    issueType = 'incorrect_info';
    subIssue = 'Account information incorrect';
  } else if (hasInvestigationIssue) {
    issueType = 'problem_with_investigation';
    subIssue = 'Their investigation did not fix an error on your report';
  }

  return {
    productType,
    issueType,
    subIssue,
    companyName: fields.furnisherOrCollector || fields.originalCreditor || '[COMPANY NAME]',
    consumerName: consumerInfo.name,
    consumerAddress: consumerInfo.address,
    consumerCity: consumerInfo.city,
    consumerState: consumerInfo.state,
    consumerZip: consumerInfo.zip,
    consumerEmail: consumerInfo.email,
    consumerPhone: consumerInfo.phone,
    narrative: generateCFPBNarrative(fields, flags),
    desiredResolution: generateDesiredResolution(flags),
    submittedToCompany: true,
    serviceMemberStatus: false,
    consentToPublish: 'with_consent',
    hasAttachments: true,
    attachmentDescriptions: generateAttachmentList(flags)
  };
}

/**
 * Generate comprehensive CFPB narrative
 */
export function generateCFPBNarrative(fields: CreditFields, flags: RuleFlag[]): string {
  const lines: string[] = [];

  // Opening
  lines.push(`I am filing this complaint regarding inaccurate information being reported on my credit file by ${fields.furnisherOrCollector || fields.originalCreditor || 'the furnisher'}.`);
  lines.push('');

  // Account identification
  lines.push('ACCOUNT INFORMATION:');
  lines.push(`- Original Creditor: ${fields.originalCreditor || 'Not provided'}`);
  lines.push(`- Current Furnisher: ${fields.furnisherOrCollector || 'Not provided'}`);
  lines.push(`- Account Type: ${fields.accountType || 'Not specified'}`);
  lines.push(`- Reported Balance: ${fields.currentBalance || 'Not provided'}`);
  lines.push(`- Date of First Delinquency: ${fields.dofd || 'Not provided'}`);
  lines.push(`- Estimated Removal Date: ${fields.estimatedRemovalDate || 'Not provided'}`);
  lines.push('');

  // Timeline of events
  lines.push('TIMELINE OF EVENTS:');
  if (fields.dateOpened) lines.push(`- Account opened: ${fields.dateOpened}`);
  if (fields.dofd) lines.push(`- First delinquency: ${fields.dofd}`);
  if (fields.chargeOffDate) lines.push(`- Charge-off date: ${fields.chargeOffDate}`);
  if (fields.dateLastPayment) lines.push(`- Last payment: ${fields.dateLastPayment}`);
  lines.push('');

  // Specific violations
  lines.push('SPECIFIC INACCURACIES AND VIOLATIONS:');
  lines.push('');

  const highFlags = flags.filter(f => f.severity === 'high');
  const mediumFlags = flags.filter(f => f.severity === 'medium');

  highFlags.forEach((flag, i) => {
    lines.push(`${i + 1}. ${flag.ruleName.toUpperCase()}`);
    lines.push(`   Issue: ${flag.explanation}`);
    lines.push(`   Why this matters: ${flag.whyItMatters}`);
    lines.push(`   Legal basis: ${flag.legalCitations.join(', ')}`);
    lines.push('');
  });

  if (mediumFlags.length > 0) {
    lines.push('ADDITIONAL CONCERNS:');
    mediumFlags.forEach(flag => {
      lines.push(`- ${flag.ruleName}: ${flag.explanation}`);
    });
    lines.push('');
  }

  // Previous attempts
  lines.push('PREVIOUS ATTEMPTS TO RESOLVE:');
  lines.push('I have previously disputed this information directly with the credit bureaus and/or the furnisher. Despite my disputes, the inaccurate information continues to be reported.');
  lines.push('');

  // Impact statement
  lines.push('IMPACT ON CONSUMER:');
  lines.push('This inaccurate reporting has negatively impacted my:');
  lines.push('- Credit score and creditworthiness');
  lines.push('- Ability to obtain credit at favorable terms');
  lines.push('- Housing opportunities (rental applications)');
  lines.push('- Employment opportunities (credit checks)');
  lines.push('- Overall financial well-being and stress levels');
  lines.push('');

  // Legal references
  lines.push('APPLICABLE LAW:');
  const citations = new Set<string>();
  flags.forEach(f => f.legalCitations.forEach(c => citations.add(c)));
  Array.from(citations).forEach(cite => {
    lines.push(`- ${cite}`);
  });
  lines.push('');

  // Closing
  lines.push('I request that the CFPB investigate this matter and take appropriate action to ensure the furnisher corrects or deletes the inaccurate information from my credit file.');

  return lines.join('\n');
}

/**
 * Generate desired resolution based on violations
 */
function generateDesiredResolution(flags: RuleFlag[]): string {
  const resolutions: string[] = [];

  resolutions.push('I am seeking the following resolution:');
  resolutions.push('');
  resolutions.push('1. Complete deletion of this inaccurate tradeline from all three credit bureaus (Experian, Equifax, TransUnion)');
  resolutions.push('');
  resolutions.push('2. Alternatively, correction of the following specific items:');

  if (flags.some(f => ['B1', 'B2', 'B3'].includes(f.ruleId))) {
    resolutions.push('   - Correction of the Date of First Delinquency to the accurate date');
    resolutions.push('   - Correction of the estimated removal date to 7 years from the true DOFD');
  }

  if (flags.some(f => f.ruleId === 'D1')) {
    resolutions.push('   - Correction of the account status to reflect paid/closed status');
    resolutions.push('   - Correction of the balance to $0');
  }

  if (flags.some(f => ['K1', 'K7'].includes(f.ruleId))) {
    resolutions.push('   - Correction of the reported balance to reflect the accurate amount');
  }

  resolutions.push('');
  resolutions.push('3. Written confirmation from the furnisher that they have corrected their records');
  resolutions.push('');
  resolutions.push('4. Written confirmation from each credit bureau that the information has been corrected or deleted');
  resolutions.push('');
  resolutions.push('5. Investigation into the furnisher\'s credit reporting practices to prevent future violations');

  return resolutions.join('\n');
}

/**
 * Generate attachment list
 */
function generateAttachmentList(flags: RuleFlag[]): string[] {
  const attachments: string[] = [
    'Copy of credit report showing disputed information',
    'Copy of government-issued ID',
    'Proof of current address'
  ];

  if (flags.some(f => ['B1', 'B2', 'B3'].includes(f.ruleId))) {
    attachments.push('Historical credit reports showing date changes');
    attachments.push('Timeline analysis showing re-aging evidence');
  }

  if (flags.some(f => f.ruleId === 'D1')) {
    attachments.push('Proof of payment (if applicable)');
    attachments.push('Settlement agreement (if applicable)');
  }

  attachments.push('Previous dispute letters sent');
  attachments.push('Response letters received (if any)');
  attachments.push('Certified mail receipts');

  return attachments;
}

/**
 * Format complaint for CFPB submission form
 */
export function formatForCFPBSubmission(complaint: CFPBComplaint): string {
  const lines: string[] = [];

  lines.push('='.repeat(60));
  lines.push('CFPB COMPLAINT - READY FOR SUBMISSION');
  lines.push('='.repeat(60));
  lines.push('');
  lines.push('NOTE: Copy each section into the corresponding field on');
  lines.push('https://www.consumerfinance.gov/complaint/');
  lines.push('');
  lines.push('-'.repeat(60));
  lines.push('PRODUCT TYPE: ' + complaint.productType.replace('_', ' ').toUpperCase());
  lines.push('-'.repeat(60));
  lines.push('');
  lines.push('-'.repeat(60));
  lines.push('ISSUE TYPE: ' + complaint.issueType.replace(/_/g, ' ').toUpperCase());
  if (complaint.subIssue) {
    lines.push('SUB-ISSUE: ' + complaint.subIssue);
  }
  lines.push('-'.repeat(60));
  lines.push('');
  lines.push('-'.repeat(60));
  lines.push('COMPANY NAME: ' + complaint.companyName);
  lines.push('-'.repeat(60));
  lines.push('');
  lines.push('-'.repeat(60));
  lines.push('COMPLAINT NARRATIVE:');
  lines.push('-'.repeat(60));
  lines.push(complaint.narrative);
  lines.push('');
  lines.push('-'.repeat(60));
  lines.push('DESIRED RESOLUTION:');
  lines.push('-'.repeat(60));
  lines.push(complaint.desiredResolution);
  lines.push('');
  lines.push('-'.repeat(60));
  lines.push('ATTACHMENTS TO INCLUDE:');
  lines.push('-'.repeat(60));
  complaint.attachmentDescriptions.forEach((att, i) => {
    lines.push(`${i + 1}. ${att}`);
  });
  lines.push('');
  lines.push('='.repeat(60));
  lines.push('CONSUMER INFORMATION (keep private)');
  lines.push('='.repeat(60));
  lines.push(`Name: ${complaint.consumerName}`);
  lines.push(`Address: ${complaint.consumerAddress}`);
  lines.push(`City: ${complaint.consumerCity}`);
  lines.push(`State: ${complaint.consumerState}`);
  lines.push(`ZIP: ${complaint.consumerZip}`);
  if (complaint.consumerEmail) lines.push(`Email: ${complaint.consumerEmail}`);
  if (complaint.consumerPhone) lines.push(`Phone: ${complaint.consumerPhone}`);

  return lines.join('\n');
}

/**
 * Get CFPB complaint URL
 */
export function getCFPBSubmissionURL(): string {
  return 'https://www.consumerfinance.gov/complaint/';
}

/**
 * Generate one-click CFPB complaint data for pre-filled form
 */
export function generateQuickComplaintData(
  fields: CreditFields,
  flags: RuleFlag[]
): {
  formUrl: string;
  summary: string;
  keyPoints: string[];
  estimatedTime: string;
} {
  const hasReaging = flags.some(f => ['B1', 'B2', 'B3', 'K6'].includes(f.ruleId));
  const creditor = fields.furnisherOrCollector || fields.originalCreditor || 'Unknown';

  const keyPoints: string[] = [];

  if (hasReaging) {
    keyPoints.push('Illegal debt re-aging detected (date manipulation)');
  }

  const highFlags = flags.filter(f => f.severity === 'high');
  if (highFlags.length > 0) {
    keyPoints.push(`${highFlags.length} high-severity FCRA violations`);
  }

  if (flags.some(f => f.ruleId === 'K6')) {
    keyPoints.push('Account reported beyond 7-year legal limit');
  }

  if (flags.some(f => f.legalCitations.some(c => c.includes('FDCPA')))) {
    keyPoints.push('FDCPA (debt collection) violations');
  }

  return {
    formUrl: 'https://www.consumerfinance.gov/complaint/',
    summary: `File complaint against ${creditor} for ${flags.length} documented violations`,
    keyPoints,
    estimatedTime: '10-15 minutes with prepared documentation',
  };
}

/**
 * Track complaint status suggestions
 */
export function getComplaintFollowUpSteps(): string[] {
  return [
    '1. Save your complaint confirmation number',
    '2. Check CFPB portal for company response (usually 15 days)',
    '3. Review and dispute company response if inadequate',
    '4. Request CFPB investigation if response is unsatisfactory',
    '5. Consider small claims court if no resolution after 60 days',
    '6. Consult with FCRA attorney for potential lawsuit if damages exceed $1000',
  ];
}

/**
 * Estimate complaint strength
 */
export function estimateComplaintStrength(flags: RuleFlag[]): {
  strength: 'strong' | 'moderate' | 'weak';
  score: number;
  factors: string[];
} {
  let score = 50; // Base score
  const factors: string[] = [];

  // High severity violations
  const highCount = flags.filter(f => f.severity === 'high').length;
  if (highCount >= 3) {
    score += 30;
    factors.push(`${highCount} high-severity violations documented`);
  } else if (highCount >= 1) {
    score += 15;
    factors.push(`${highCount} high-severity violation(s) documented`);
  }

  // Re-aging is particularly strong
  if (flags.some(f => ['B1', 'B2', 'B3'].includes(f.ruleId))) {
    score += 20;
    factors.push('Clear evidence of date manipulation (re-aging)');
  }

  // Multiple legal citations
  const citations = new Set<string>();
  flags.forEach(f => f.legalCitations.forEach(c => citations.add(c)));
  if (citations.size >= 3) {
    score += 10;
    factors.push(`Multiple legal violations cited (${citations.size} statutes)`);
  }

  // Timeline issues
  if (flags.some(f => f.ruleId === 'K6')) {
    score += 15;
    factors.push('Account exceeds 7-year reporting limit');
  }

  // Cap at 100
  score = Math.min(100, score);

  return {
    strength: score >= 75 ? 'strong' : score >= 50 ? 'moderate' : 'weak',
    score,
    factors
  };
}
